guice-jpa/src/main/java/com/maxifier/guice/jpa/DBInterceptor.java

```diff
@@ -1,342 +1,98 @@
-
-    /**
-     * Cache for EMFs
-     */
-    private final Map<Object, EntityManagerFactory> emfsCache = new HashMap<Object, EntityManagerFactory>();
-
-    /**
-     * Method cache for annotations.
-     * Annotation[] array: [0] - @DB, [1] - binding Annotation.
-     */
-    private final Map<Method, Annotation[]> methodsCache = new WeakHashMap<Method, Annotation[]>();
-
-    /**
-     * List of used with annotations needed to be checked.
-     */
-    private final List<Annotation[]> awaiting = new LinkedList<Annotation[]>();
-
-    /**
-     * Does this interceptor prepared to work?
-     */
-    private volatile boolean prepared = false;
-
-    /**
-     * This is delegate proxy for {@link EntityManager} calls.
-     */
-    @SuppressWarnings("EntityManagerInspection")
-    private EntityManager wrapper;
-
-    @Inject
-    synchronized void prepare(Injector injector) {
-        //find all EMFs
-        List<Binding<EntityManagerFactory>> bindingsByType
-                = injector.findBindingsByType(TypeLiteral.get(EntityManagerFactory.class));
-        //prepare emfs cache and contexts
-        for (Binding<EntityManagerFactory> binding : bindingsByType) {
-            Key<EntityManagerFactory> key = binding.getKey();
-            if (key.getAnnotation() != null) {
-                emfsCache.put(key.getAnnotation(), binding.getProvider().get());
-                contexts.put(key.getAnnotation(), new ThreadLocal<EntityManager>());
-            } else if (key.getAnnotationType() != null) {
-                emfsCache.put(key.getAnnotationType(), binding.getProvider().get());
-                contexts.put(key.getAnnotationType(), new ThreadLocal<EntityManager>());
-            } else {
-                //default binding
-                emfsCache.put(DEFAULT, binding.getProvider().get());
-                contexts.put(DEFAULT, new ThreadLocal<EntityManager>());
-            }
-        }
-        //create wrapper
-        Enhancer enhancer = new Enhancer();
-        enhancer.setInterfaces(new Class[]{EntityManager.class});
-        enhancer.setCallback(new net.sf.cglib.proxy.MethodInterceptor() {
-            @Override
-            public Object intercept(Object obj, Method method, Object[] args, MethodProxy proxy) throws Throwable {
-                if (method.getName().equals("close")) {
-                    //do nothing
-                    return null;
-                } else {
-                    return proxy.invoke(current(), args);
-                }
-
-            }
-        });
-        wrapper = (EntityManager) enhancer.create();
-        prepared = true;
-    }
-
+public class DBInterceptor implements MethodInterceptor {
+    private static final Logger logger = LoggerFactory.getLogger(DBInterceptor.class);
+    private static final Random RND = new Random();
+    private static final int RETRY_TIMEOUTS[] = new int[]{
+        0, 307, 2000, 11000, 19000, 31000, 53000, 89000, 151000, 241000, 307000
```

Comment:
Oops I thought there's one in Guava, but it is actually situated in DFP API lib...